asio/include/asio/detail/impl/epoll_reactor.ipp

```diff
@@ -503,6 +503,24 @@ void epoll_reactor::run(long usec, op_queue<operation>& ops)
 #if defined(ASIO_HAS_TIMERFD)
     else if (ptr == &timer_fd_)
     {
+      // We should read an 8-byte timeout expiration counter (using
+      // the "timerfd" descriptor) every time when the EPOLLIN event
+      // occurs. Otherwise, the timer stops to notifying the application
+      // about new events, and all activity associated with this timer
+      // will be frozen on some systems, or vice versa, we get an infinite
+      // loop of notifications because of the constant readiness to read
+      // the new data:
+
+      if (events[i].events & EPOLLIN)
+      {
+        // If timeout was expired, we should read the expiration counter:
+        uint64_t count;
+        if (read(timer_fd_, &count, sizeof(count)))
```

Comment:
After returning from the epoll_wait() with EPOLLIN notification the read() system call is never blocked, so it never returns EINTR in the real world. And, since timerfd_create() returns pseudo-device handle (instead of the real device), then read() never return less than 8 bytes. But formally, for reassurance, we can write some code like this:
```
ssize_t len = sizeof(count);
char * buf = reinterpret_cast<char *>(&count);
do {
  ssize_t ret = read(timer_fd_, buf, len);
  if (ret < 0)
  {
    If (EINTR != errno) break;
    continue;
  }
  buf += ret;
  len -= ret;
} while (len);
```